cctp_primitives/src/utils/commitment_tree.rs

```diff
@@ -1,71 +1,50 @@
-use primitives::{
-    FieldBasedHash, FieldBasedMerkleTree,
-    merkle_tree::field_based_mht::parameters::tweedle_fr::TWEEDLE_MHT_POSEIDON_PARAMETERS as MHT_PARAMETERS
-};
-use crate::commitment_tree::{FieldElement, FieldHash, FieldElementsMT, FIELD_SIZE};
+use primitives::FieldBasedHash;
+use crate::type_mapping::{FIELD_SIZE, FieldElement, FieldHash, GingerMHT, Error};
 use rand::Rng;
-use algebra::{to_bytes, ToBytes, UniformRand, ToConstraintField};
-
-pub type Error = Box<dyn std::error::Error>;
+use algebra::{UniformRand, ToConstraintField, CanonicalSerialize};
+use crate::utils::mht::{new_ginger_mht, append_leaf_to_ginger_mht};
 
 pub const fn pow2(power: usize) -> usize { 1 << power }
 
 //--------------------------------------------------------------------------------------------------
 // Merkle Tree utils
 //--------------------------------------------------------------------------------------------------
 
-// Creates new FieldElement-based MT
-pub fn new_mt(height: usize) -> Result<FieldElementsMT, Error> {
-    let processing_step = 2usize.pow(height as u32);
-    Ok(FieldElementsMT::init(
-        height,
-        processing_step
-    ))
+/// Creates new FieldElement-based MT
+pub fn new_mt(height: usize) -> GingerMHT {
+    new_ginger_mht(height, 2usize.pow(height as u32))
 }
 
-// Sequentially inserts leafs into an MT by using a specified position which is incremented afterwards
-// Returns false if there is no more place to insert a leaf
-pub fn add_leaf(tree: &mut FieldElementsMT, leaf: &FieldElement, pos: &mut usize, capacity: usize) -> bool {
-    if *pos < capacity {
-        tree.append(*leaf); *pos += 1;
+/// Sequentially inserts leafs into an MT by using a specified position which is incremented afterwards
+/// Returns false if there is no more place to insert a leaf
+pub fn add_leaf(tree: &mut GingerMHT, leaf: &FieldElement) -> bool {
+    if append_leaf_to_ginger_mht(tree, leaf).is_ok() {
```

Comment:
The `mht` instance will return `failure` if it's full, right? So no sense to check capacity limits outside.

Reply:
Exactly